Automattic/jetpack
# AI Assistant: Use backend system prompt

Part of #28569.

## Proposed changes:

* Start using backend system prompts
* This change uses a `role=jetpack-ai` message to request a generated system prompt from the backend
* The message being sent will be transformed to a system message and a relevant content message, if applicable
* The format of the message is:

```
{
   role: 'jetpack-ai',
   context: {
      type: 'ai-assistant-initial-prompt',
      for: 'the prompt type to which the system prompt will be used, since the type can modify the system prompt'
      content?: 'relevant content for the request, when applicable',
      custom_system_prompt?: 'the play ground system prompt; when set, will override the backend version'
   }
}
```

* The function that generates the message is replacing the current mechanism of creating the prompt but I did not remove the code yet, since we will need a whole clean-up to remove all the dead flows
* This change focus on the system prompts so it's more simple to test and check just this change instead of changing the whole app; follow-up PRs will start enabling backend prompts for the user actions as well
* The code contains some logic to generate user prompt messages as well, but it's not enabled yet and we will have the opportunity to test it on the next PRs

Some lateral changes made:

* Removed the "fake user-assistant" history that was used to set the content when using the block extension, since now the initial message will take care of creating the content history on the backend
* Removed the `contentType: generated` set when transforming extended blocks to the assistant block, since the `generatedContent` value is a source more valid given the new way of handling the relevant content on the backend

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






* ~~Test in pair with D121131-code~~ backend change got deployed already
* ~~Make sure your test site is pointing to the `public-api`in your sandbox, and your sandbox is running the code above~~
* Go to the block editor and open the browser console
* Enable debug messages by running the following command; refresh the window after it:

```
localStorage.setItem( 'debug', '*' );
```

* Insert an AI Assistant block and ask for some completion (example: a haiku about code)
* Notice the `jetpack-ai-client:ask-question` log message
   * the message will follow the new format
   * if the post has some content, confirm the correct relevant content is present on the message `context.content` property
   * confirm that the `context.for` property contains the right target action

<img width="800" alt="Screenshot 2023-09-08 at 18 54 14" src="https://github.com/Automattic/jetpack/assets/6760046/fe4ee878-8066-4910-b864-75283fc54f49">

* Confirm there is no `role=system` message
* Confirm the request works as expected:

<img width="700" alt="Screenshot 2023-09-08 at 18 56 33" src="https://github.com/Automattic/jetpack/assets/6760046/b3f439a2-8b88-4767-8e23-1837a07ed5fa">

* Confirm the same happens for all the other actions (simplify, change tone, translate, etc)
* Make sure we don't have regressions, and that you can keep iterating over the content and the `ai-assistant-initial-message` is always present:

<img width="800" alt="Screenshot 2023-09-08 at 18 58 06" src="https://github.com/Automattic/jetpack/assets/6760046/3d2cd2b9-9e1c-4cea-b93f-4bc51e08dd94">

* Use the logstash logs to confirm that the prompt being executed on the OpenAI API is correct
* Do the same logtash log check for actions that would modify the system prompt, like the "Simplify" that would add three more rules to it:

<img width="600" alt="Screenshot 2023-09-08 at 18 51 42" src="https://github.com/Automattic/jetpack/assets/6760046/28f84749-789f-4313-9fd4-7590bf19d741">

<img width="800" alt="Screenshot 2023-09-08 at 19 01 27" src="https://github.com/Automattic/jetpack/assets/6760046/8cbe1a86-7a93-49d6-8e22-30b314af3fb5">

<img width="800" alt="Screenshot 2023-09-08 at 19 02 24" src="https://github.com/Automattic/jetpack/assets/6760046/6f2cf2df-874a-4560-834b-3015de964731">

* Confirm that you can set a custom system prompt on the client playground and that custom prompt will take precedence over the backend version:

<img width="800" alt="Screenshot 2023-09-08 at 19 03 58" src="https://github.com/Automattic/jetpack/assets/6760046/51cc7107-9fcb-447b-abc3-5769d81426f8">

<img width="800" alt="Screenshot 2023-09-08 at 19 04 45" src="https://github.com/Automattic/jetpack/assets/6760046/47c1b3c2-481f-4fd7-9e47-8e8d1ae7cfc8">

<img width="800" alt="Screenshot 2023-09-08 at 19 05 19" src="https://github.com/Automattic/jetpack/assets/6760046/ce428ce6-cc07-40cb-a471-641132749a49">
